### ISSUE TITLE
Unable to make TCP DNS queries on one cluster, not others, until DestinationRule recreated

### ISSUE THREAD

**Issue by Stono** [istio/istio#11658]
Hi, 
We have three clusters, dev, preprod and prod, which are all built identically (same Infra as Code etc).  On one of them (preprod), applications that are behind Istio are unable to make TCP DNS queries:

```
[atcloud@airflow-web-5d9984d64-2pqpm app]$ dig +tcp www.google.com
;; communications error to 10.192.32.10#53: connection reset
```

Looking at istio debug logs, there's nothing particularly telling:
```
[2019-02-11 08:07:38.552][28][debug][filter] external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[2019-02-11 08:07:38.552][28][debug][filter] src/envoy/tcp/mixer/filter.cc:28] Called tcp filter: Filter
[2019-02-11 08:07:38.552][28][debug][filter] src/envoy/tcp/mixer/filter.cc:38] Called tcp filter: initializeReadFilterCallbacks
[2019-02-11 08:07:38.552][28][debug][filter] external/envoy/source/common/tcp_proxy/tcp_proxy.cc:168] [C423] new tcp proxy session
[2019-02-11 08:07:38.552][28][debug][filter] src/envoy/tcp/mixer/filter.cc:98] [C423] Called tcp filter onNewConnection: remote 10.198.13.22:49261, local 10.192.32.10:53
[2019-02-11 08:07:38.552][28][debug][filter] external/envoy/source/common/tcp_proxy/tcp_proxy.cc:305] [C423] Creating connection to cluster outbound|53||kube-dns.kube-system.svc.cluster.local
[2019-02-11 08:07:38.552][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:98] [C423] closing data_to_write=0 type=1
[2019-02-11 08:07:38.552][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:133] [C423] closing socket: 1
[2019-02-11 08:07:38.552][28][debug][filter] src/envoy/tcp/mixer/filter.cc:134] Called tcp filter onEvent: 1
[2019-02-11 08:07:38.552][28][debug][main] external/envoy/source/server/connection_handler_impl.cc:218] [C423] new connection
[2019-02-11 08:07:38.552][28][debug][filter] src/envoy/tcp/mixer/filter.cc:33] Called tcp filter : ~Filter
[2019-02-11 08:07:39.746][23][debug][main] external/envoy/source/server/server.cc:126] flushing stats
```

Pods in Dev and Prod are absolutely fine.

We have mTLS enabled on our cluster, and have done since day one.  As a result we have had the following destinationrule in place for almost 3 months:

```
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  creationTimestamp: 2018-11-05T17:35:35Z
  generation: 1
  labels:
    app: istio-crd
    chart: istio-crd-1.0.5
    heritage: Tiller
    release: istio-crd
  name: disable-mtls-to-k8s-dns
  namespace: istio-system
  resourceVersion: "80583893"
  selfLink: /apis/networking.istio.io/v1alpha3/namespaces/istio-system/destinationrules/disable-mtls-to-k8s-dns
  uid: 339f9c78-e121-11e8-a96c-42010aa40060
spec:
  host: kube-dns.kube-system.svc.cluster.local
  trafficPolicy:
    tls:
      mode: DISABLE
```

Because I enjoy shooting in the dark, and because I'd seen TCP dns queries failed in a similar fashion months ago before we put this destination rule in, I decided to delete and recreate the DestinationRule, and this **fixed the situation**:

```
[atcloud@airflow-web-5d9984d64-2pqpm app]$ dig +tcp www.google.com

; <<>> DiG 9.9.4-RedHat-9.9.4-73.el7_6 <<>> +tcp www.google.com
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 19108
;; flags: qr rd ra; QUERY: 1, ANSWER: 6, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 512
;; QUESTION SECTION:
;www.google.com.                        IN      A

;; ANSWER SECTION:
www.google.com.         299     IN      A       108.177.96.147
www.google.com.         299     IN      A       108.177.96.103
```

```
[2019-02-11 08:15:18.731][28][debug][filter] external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[2019-02-11 08:15:18.731][28][debug][filter] src/envoy/tcp/mixer/filter.cc:28] Called tcp filter: Filter
[2019-02-11 08:15:18.731][28][debug][filter] src/envoy/tcp/mixer/filter.cc:38] Called tcp filter: initializeReadFilterCallbacks
[2019-02-11 08:15:18.731][28][debug][filter] external/envoy/source/common/tcp_proxy/tcp_proxy.cc:168] [C436] new tcp proxy session
[2019-02-11 08:15:18.731][28][debug][filter] src/envoy/tcp/mixer/filter.cc:98] [C436] Called tcp filter onNewConnection: remote 10.198.13.22:57143, local 10.192.32.10:53
[2019-02-11 08:15:18.731][28][debug][filter] external/envoy/source/common/tcp_proxy/tcp_proxy.cc:305] [C436] Creating connection to cluster outbound|53||kube-dns.kube-system.svc.cluster.local
[2019-02-11 08:15:18.731][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:572] [C437] connecting to 10.198.10.9:53
[2019-02-11 08:15:18.731][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:581] [C437] connection in progress
[2019-02-11 08:15:18.731][28][debug][main] external/envoy/source/server/connection_handler_impl.cc:218] [C436] new connection
[2019-02-11 08:15:18.732][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:466] [C437] connected
[2019-02-11 08:15:18.733][28][debug][filter] src/envoy/tcp/mixer/filter.cc:105] Called tcp filter completeCheck: OK
[2019-02-11 08:15:18.733][28][debug][filter] src/envoy/tcp/mixer/filter.cc:78] [C436] Called tcp filter onRead bytes: 45
[2019-02-11 08:15:18.737][28][debug][filter] src/envoy/tcp/mixer/filter.cc:88] [C436] Called tcp filter onWrite bytes: 141
[2019-02-11 08:15:18.737][28][debug][filter] src/envoy/tcp/mixer/filter.cc:78] [C436] Called tcp filter onRead bytes: 0
[2019-02-11 08:15:18.738][28][debug][filter] src/envoy/tcp/mixer/filter.cc:88] [C436] Called tcp filter onWrite bytes: 0
[2019-02-11 08:15:18.738][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:451] [C437] remote close
[2019-02-11 08:15:18.738][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:133] [C437] closing socket: 0
[2019-02-11 08:15:18.738][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:98] [C436] closing data_to_write=0 type=0
[2019-02-11 08:15:18.738][28][debug][connection] external/envoy/source/common/network/connection_impl.cc:133] [C436] closing socket: 1
[2019-02-11 08:15:18.738][28][debug][filter] src/envoy/tcp/mixer/filter.cc:132] Called tcp filter onEvent: 1 upstream 10.198.10.9:53
[2019-02-11 08:15:18.738][28][debug][main] external/envoy/source/server/connection_handler_impl.cc:51] [C436] adding to cleanup list
[2019-02-11 08:15:18.738][28][debug][filter] src/envoy/tcp/mixer/filter.cc:33] Called tcp filter : ~Filter
```

Really worried about this, that a DestinationRule appears to have been randomly ignored?  It's worth noting that all other DestinationRules appear to be being honoured.

**Comment by Stono** [comment]
OK 
This came back only moments later.  This time we tried restarting pilot and again this has (probably temporarily) fixed the problem.

**I have captured cluster debug logs from both before and after** - please can someone from istio reach out to me directly on Slack so I can provide them securely.

Thanks
Karl

**Comment by Stono** [comment]
It has returned after an istio-pilot restart too.

**Comment by Stono** [comment]
OK - latest changes, I set `PILOT_DIRECT_EDS=0` to disable the incremental EDS updates because the only difference between these clusters is that preprod has *more pods* than the other clusters.

Setting `PILOT_DIRECT_EDS=0` seems to have resolved the problem and our DNS queries are now working fine and dandy again.

cc @costinm

**Comment by Stono** [comment]
@costinm @rshriram we turned `PILOT_DIRECT_EDS` back on today after upgrading to 1.0.6 and this error still persists. 

It appears that istio-proxies do not have any endpoints on the kubedns cluster so TCP dns requests fail: 
```
[atcloud@sauron-web-5b5df8f647-2v6q7 app]$ curl -s http://127.0.0.1:15000/clusters | grep kube-dns
outbound|53||kube-dns.kube-system.svc.cluster.local::default_priority::max_connections::1024
outbound|53||kube-dns.kube-system.svc.cluster.local::default_priority::max_pending_requests::1024
outbound|53||kube-dns.kube-system.svc.cluster.local::default_priority::max_requests::1024
outbound|53||kube-dns.kube-system.svc.cluster.local::default_priority::max_retries::3
outbound|53||kube-dns.kube-system.svc.cluster.local::high_priority::max_connections::1024
outbound|53||kube-dns.kube-system.svc.cluster.local::high_priority::max_pending_requests::1024
outbound|53||kube-dns.kube-system.svc.cluster.local::high_priority::max_requests::1024
outbound|53||kube-dns.kube-system.svc.cluster.local::high_priority::max_retries::3
outbound|53||kube-dns.kube-system.svc.cluster.local::added_via_api::true
```

The service definitely has endpoints:
```
❯ kks get ep kube-dns -o yaml
apiVersion: v1
kind: Endpoints
metadata:
  creationTimestamp: 2018-05-11T16:02:05Z
  labels:
    addonmanager.kubernetes.io/mode: Reconcile
    k8s-app: kube-dns
    kubernetes.io/cluster-service: "true"
    kubernetes.io/name: KubeDNS
  name: kube-dns
  namespace: kube-system
  resourceVersion: "113798791"
  selfLink: /api/v1/namespaces/kube-system/endpoints/kube-dns
  uid: a6b930b9-5534-11e8-9910-42010aa400bf
subsets:
- addresses:
  - ip: 10.198.13.109
    nodeName: gke-delivery-platfor-durable-20190114-d2a699bd-zd3k
    targetRef:
      kind: Pod
      name: kube-dns-7549f99fcc-jg2vw
      namespace: kube-system
      resourceVersion: "113798336"
      uid: 11d04e44-345a-11e9-9430-42010aa40128
  - ip: 10.198.13.110
    nodeName: gke-delivery-platfor-durable-20190114-d2a699bd-zd3k
    targetRef:
      kind: Pod
      name: kube-dns-7549f99fcc-q4mrt
      namespace: kube-system
      resourceVersion: "113798544"
      uid: 17c2d9fd-345a-11e9-9430-42010aa40128
  - ip: 10.198.18.51
    nodeName: gke-delivery-platfor-durable-20190114-ea1efef2-n4w4
    targetRef:
      kind: Pod
      name: kube-dns-7549f99fcc-fq2pk
      namespace: kube-system
      resourceVersion: "113798740"
      uid: 1f535b05-345a-11e9-9430-42010aa40128
  - ip: 10.198.9.125
    nodeName: gke-delivery-platfor-durable-20190114-4d15f0a5-s7bz
    targetRef:
      kind: Pod
      name: kube-dns-7549f99fcc-hnrls
      namespace: kube-system
      resourceVersion: "113798782"
      uid: 27229fd5-345a-11e9-9430-42010aa40128
  ports:
  - name: dns
    port: 53
    protocol: UDP
  - name: dns-tcp
    port: 53
    protocol: TCP
```

And pilot is aware of them (my colleague is going to post the pilot config in a second).  

The *only happens on our largest cluster*, which is leading me to believe that it is potentially something to do with the number of endpoints/services or something similar but there is nothing telling in the logs, other than the pod detection:

```
❯ kis logs istio-pilot-5868cc6688-g2d25 discovery | grep kube-dns
2019-02-20T05:00:14.146913Z     info    Handling event add for pod kube-dns-7549f99fcc-fq2pk in namespace kube-system -> 10.198.18.51
2019-02-20T05:00:14.149150Z     info    Handling event add for pod kube-dns-autoscaler-67c97c87fb-l2g22 in namespace kube-system -> 10.198.13.21
2019-02-20T05:00:14.149737Z     info    Handling event add for pod kube-dns-7549f99fcc-hnrls in namespace kube-system -> 10.198.9.125
2019-02-20T05:00:14.151760Z     info    Handling event add for pod kube-dns-7549f99fcc-q4mrt in namespace kube-system -> 10.198.13.110
2019-02-20T05:00:14.151907Z     info    Handling event add for pod kube-dns-7549f99fcc-jg2vw in namespace kube-system -> 10.198.13.109
```

So for whatever reason, that config is not getting to the side cars.

**Comment by chrismckean** [comment]
{"svc": "kube-dns.kube-system.svc.cluster.local:dns", "ep": [
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-jg2vw.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-a",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-q4mrt.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-a",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-fq2pk.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-b",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-hnrls.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-c",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },

{}]},
{"svc": "kube-dns.kube-system.svc.cluster.local:dns-tcp", "ep": [
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-jg2vw.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-a",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-q4mrt.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-a",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-fq2pk.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-b",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },
{
    "endpoint": {
      "Family": 0,
      "Address": "10.x.x.x",
      "Port": 53,
      "ServicePort": {
        "name": "dns",
        "port": 53,
        "protocol": "UDP"
      },
      "UID": "kubernetes://kube-dns-7549f99fcc-hnrls.kube-system"
    },
    "service": {
      "hostname": "kube-dns.kube-system.svc.cluster.local",
      "address": "10.x.x.x",
      "ports": [
        {
          "name": "dns",
          "port": 53,
          "protocol": "UDP"
        },
        {
          "name": "dns-tcp",
          "port": 53,
          "protocol": "TCP"
        }
      ],
      "external": "",
      "MeshExternal": false,
      "Resolution": 0,
      "creationTime": "2018-05-11T16:02:05Z",
      "Attributes": {
        "Name": "kube-dns",
        "Namespace": "kube-system",
        "UID": "istio://kube-system/services/kube-dns"
      }
    },
    "labels": {
      "k8s-app": "kube-dns",
      "pod-template-hash": "3105955977"
    },
    "az": "europe-west4/europe-west4-c",
    "serviceaccount": "spiffe://cluster.local/ns/kube-system/sa/kube-dns"
  },

**Comment by costinm** [comment]
So pilot has the endpoints, a DR change (which triggers full push) fixes it. And the endpoints are stable - so somehow when the connection happened the cluster was not sent. 

I don't think the 'incremental' part is the root cause - disabling incremental just makes it that any change of endpoint anywhere in the cluster is doing a full push, hiding the problem.

**Comment by costinm** [comment]
Current status: it seems the root problem is envoy not getting some endpoints - the CDS cluster is sent ( based on the stats ), Pilot has the endpoints in memory - and it appears that a full push will send them properly.  Both the change in DR, pilot restart or turning off incremental EDS ( which results in frequent updates ) are recovering. So a full push updates - but somehow in some cases it gets an empty list for this specific cluster. 

Unfortunately the logs I got have the ADS logging turned off to warning - there are some messages showing what Pilot is sending (stats and timing of the pushes). It would help a lot to repro with this enabled.

Second info would be an envoy config_dump and full /stats from the affected sidecar, and in general additional envoy logs from that job. 

Code review: so far nothing interesting, the full push is recomputing clusters, and we send the cached value. But I see no code path where the clusters will end up empty - kubedns is pretty stable.

**Comment by costinm** [comment]
Found the problem, kube-system is actually excluded by code.

Needs to be fixed in 1.0 and 1.1 branches - seems like a release blocker, but easy to fix (since calls to apiserver and dns can both fail in unexpected ways)

**Comment by rshriram** [comment]
@Stono can you create a service Entry for kube-system with resolution DNS ? I think that should solve the problem

**Comment by Stono** [comment]
I can, but, I don't understand why this is a problem on one cluster, but not the other two, when they're **identical**, other than the number of services deployed.

If it was something as deterministic as what you're both saying, i'd see consistently across clusters?

**Comment by Stono** [comment]
@costinm are you linking a PR to this issue? 
Wouldn't want this leaking into 1.1 as it'd make 1.1 unusable for us

**Comment by munrodg** [comment]
@duderino need to create a 1.0.7 with this fix.

**Comment by duderino** [comment]
1.0.7:  https://github.com/istio/istio/issues/12090

**Comment by duderino** [comment]
@howardjohn @costinm @munrodg @Stono @utako I think the current state is we're good for 1.1 but still waiting on a fix on the release-1.0 branch before we cut 1.0.7

**Comment by howardjohn** [comment]
Yes, this should be fixed and merged in 1.1 now. Please let me know if you get similar issues in 1.1. Fix for 1.0 is not yet ready

**Comment by Stono** [comment]
What's the situation with a 1.0.7 release?  We've had to turn of incremental eds because of this which creates a lot more push lag for us on our cluster.

**Comment by Stono** [comment]
@rshriram what do you mean a ServiceEntry for `kube-system` as that's a namespace?
Can you please show me?

**Comment by howardjohn** [comment]
https://github.com/istio/istio/pull/12030 has the change for 1.0.7 if you want to track it

**Comment by Stono** [comment]
This is now a critical problem on 1.1.0, TCP DNS is broken for us and we can't set the EDS flag to fix it.

**Comment by Stono** [comment]
To be clear what we see here:

 - New pod comes up, can make TCP dns queries 
 - Proxy receives a partial config update
 - TCP dns queries fail 

Previously, we set `PILOT_DIRECT_EDS=0` to "fix" this, but that isn't an option in 1.1.0.

**Comment by Stono** [comment]
If i edit the `kube-dns` service, and add a random annotation, the other pods TCP DNS start working for a short period.  So seems like that service being edited pushes out an update which containers kube-dns.

Then another update will come along, and remove it.

**Comment by Stono** [comment]
It's worth repeating that i only see this on large clusters.  This one is 553 pods with 387 services.

**Comment by Stono** [comment]
I'm chatting to @dwradcliffe - he has a cluster of a similar size, (but about 1/4 the number of services) and doesn't see it.  He runs galley, I don't.

**Comment by duderino** [comment]
@howardjohn is working on this now

**Comment by Stono** [comment]
Keeping this up to date with slack, I can deterministically recreate this issue.

As you can see, no dns endpoints in the kube-dns cluster:
```
❯ istioctl proxy-config endpoint monkeynetes-57df648c4d-2cwrp --cluster "outbound|53||kube-dns.kube-system.svc.cluster.local"
ENDPOINT     STATUS     CLUSTER
```

And digs fails:
```
[atcloud@monkeynetes-57df648c4d-2cwrp app]$ dig +tcp @10.192.32.10 www.google.com
;; communications error to 10.192.32.10#53: connection reset
```

If i create these two resources:
```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: dns
  namespace: kube-system
spec:
  hosts:
  - "kube-dns.kube-system.svc.cluster.local"
  ports:
  - number: 53
    name: tcp
    protocol: TCP
  resolution: NONE
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: disable-mtls-to-dns
  namespace: kube-system
spec:
  host: "kube-dns.kube-system.svc.cluster.local"
  trafficPolicy:
    tls:
      mode: DISABLE

serviceentry.networking.istio.io/dns created
destinationrule.networking.istio.io/disable-mtls-to-dns created
```

I get endpoints:
```
❯ istioctl proxy-config endpoint monkeynetes-57df648c4d-2cwrp --cluster "outbound|53||kube-dns.kube-system.svc.cluster.local"
ENDPOINT            STATUS      CLUSTER
10.198.12.9:53      HEALTHY     outbound|53||kube-dns.kube-system.svc.cluster.local
10.198.28.8:53      HEALTHY     outbound|53||kube-dns.kube-system.svc.cluster.local
10.198.28.86:53     HEALTHY     outbound|53||kube-dns.kube-system.svc.cluster.local
10.198.9.62:53      HEALTHY     outbound|53||kube-dns.kube-system.svc.cluster.local
10.198.9.70:53      HEALTHY     outbound|53||kube-dns.kube-system.svc.cluster.local
```

And digs work:
```
[atcloud@monkeynetes-57df648c4d-2cwrp app]$ dig +short +tcp @10.192.32.10 www.google.com
173.194.69.105
173.194.69.104
173.194.69.103
173.194.69.106
173.194.69.99
173.194.69.147
```

If I delete those entries:
```
k serviceentry.networking.istio.io "dns" deleted
destinationrule.networking.istio.io "disable-mtls-to-dns" deleted
```

Digs continue to work:
```
[atcloud@monkeynetes-57df648c4d-2cwrp app]$ dig +short +tcp @10.192.32.10 www.google.com
173.194.69.105
173.194.69.104
173.194.69.103
173.194.69.106
173.194.69.99
173.194.69.147
```

Until I do something to trigger a partial push (deleting another pod in that namespace):
```
❯ k delete pod monkeynetes-57df648c4d-d9jvh
pod "monkeynetes-57df648c4d-d9jvh" deleted
```

Then digs start to fail:
```
[atcloud@monkeynetes-57df648c4d-2cwrp app]$ dig +short +tcp @10.192.32.10 www.google.com
;; communications error to 10.192.32.10#53: connection reset
```

And the cluster is empty again:
```
❯ istioctl proxy-config endpoint monkeynetes-57df648c4d-2cwrp --cluster "outbound|53||kube-dns.kube-system.svc.cluster.local"
ENDPOINT     STATUS     CLUSTER
```

**Comment by Stono** [comment]
For anyone else experiencing this, adding a `ServiceEntry` and a `DestinationRule` if you have mTLS enabled, works around the issue:

```
apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: dns
  namespace: kube-system
spec:
  hosts:
  - "kube-dns.kube-system.svc.cluster.local"
  ports:
  - number: 53
    name: tcp
    protocol: TCP
  resolution: NONE
---
apiVersion: networking.istio.io/v1alpha3
kind: DestinationRule
metadata:
  name: disable-mtls-to-dns
  namespace: kube-system
spec:
  host: "kube-dns.kube-system.svc.cluster.local"
  trafficPolicy:
    tls:
      mode: DISABLE
```

**Comment by philrud** [comment]
I've spent some time digging into this on Friday, sharing my findings to the date.

One thing that's special about kube-dns service is that it (being a normal DNS server) talks two protocols on the same port. I.e. its Service resource has two "service ports" (UDP and TCP) on the same physical port (53).

From reading the source code it doesn't seem like Istio supports that very well. There are several places where we perform lookups by a port number picking the first port that matches. I have setup a couple of experiments and I'm getting different behavior depending on which port occurs first in the Service resource definition. If UDP goes first (kube-dns case) I couldn't get my service to work even with ServiceEntry (kube-dns however works fine even without one). By the way ServiceEntry helps because it essentially helps to disambiguate the port. This also depends on the order of endpoints returned by K8s APIs, I'm researching whether there's any consistency to it as this might explain why this issue only affects one cluster but not the other two.

We definitely need to clean up the way we handle ports, but that isn't trivial. I'm trying to put together a straightforward/safe changeset to either add deterministic ordering or to simply skip UDP ports when we're looking up ports by a port number/iterating over them etc., still haven't got it working yet.

Regarding "incremental push breaking things after ServiceEntry is removed": with a recent change we're no longer ignoring kube-system namespace, so once ServiceEntry is gone Pilot will pick up kube-dns Service resource with conflicting ports. We need to confirm that both full push and incremental push code paths behave consistently with regards to Service vs ServiceEntry -- something may be off there.

The investigation is still ongoing but this is promising. Initially I suspected some obscure concurrency related issue that potentially may affect any service, not just kube-dns, that would've been much bigger problem, than lack of proper support of two protocols on the same port.

**Comment by Stono** [comment]
@philrud - thanks for the update.  I'm glad the blast radius of the issue seems to be a bit more confined to kube-dns.  It's just a shame that the blast radius of kube-dns is large - haha!

I do find it odd that I only get it only on the larger more complex clusters so definitely feels like something else is a bit off.  

To be clear, we never had the ServiceEntry on 1.0.6 with incremental push turned off, and never saw this problem.  So I agree with "We need to confirm that both full push and incremental push code paths behave consistently with regards to Service vs ServiceEntry".

**Comment by howardjohn** [comment]
Not the cause, but related: https://github.com/kubernetes/kubernetes/issues/39188

Bug in kubernetes causes issues with ports overlapping, could be related. Kube-dns will always have both the TCP and UDP port, but when creating own service with both, need to be careful that both are created.

**Comment by hzxuzhonghu** [comment]
@Stono Can you tell us how to reproduce, I don't think we have found the root cause.

**Comment by Stono** [comment]
@hzxuzhonghu yeah, you have :-) https://github.com/istio/istio/pull/12766/files

**Comment by duderino** [comment]
@Stono @howardjohn I think this is ready to be closed?

**Comment by howardjohn** [comment]
Yeah we can close, the fix should be in 1.1.3. Feel free to open another issue or reopen if you see this in 1.1.3 again.
